### PR TITLE
[diffusion, worker] feat: support Ascend NPU for Qwen-Image FlowGRPO …

### PR DESCRIPTION
**Pull request by panshaowu**
### What does this PR do?

This PR implements Ascend NPU platform support for FlowGRPO training, specifically targeting the diffusion model family (e.g.,
  Qwen-Image). It introduces:
   - NPU-specific memory management via CaMemAllocator to support model co-location.
   - FSDP engine adaptation for NPU with _native_npu attention backend.
   - A new DiffusionRewardLoopManager for efficient reward score assembly in multi-modal tasks.
   - Comprehensive NPU smoke tests and training examples.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `vllm_omni`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `diffusion`, `omni`, `tests`, `docker`
  - If this PR involves multiple modules, separate them with `,` like `[diffusion, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][diffusion, fsdp] feat: new rollout scheduler`

### Test

Validated on Atlas 800T A2 (CANN 9.0.0):
   1. Smoke Tests: All tests in tests/npu_smoke/run_npu_smoke_tests.sh passed.
   2. E2E Training: Successfully ran OCR LoRA training using examples/flowgrpo_trainer/run_qwen_image_ocr_lora_npu.sh with 8 NPUs (TP8 strategy), confirming stable memory usage and convergence logic.
   3. The reward value at step `0`, `30`, and `60` are `0.9052`, `0.9252`, and `0.9691`. The loss and reward curves are as follows:

image

>Note: The hyperparameters used for the above training task are bs=32, lr=3e-4. Since the developer has not yet updated the relevant shell scripts, the PR has already been merged. We will refresh the hyperparameters in #85 .

### API and Usage Example

To launch training on Ascend NPU, use the provided environment setup and script:

```shell
bash examples/flowgrpo_trainer/run_qwen_image_ocr_lora_npu.sh
```

### Design & Code Changes

- Worker & Engine:
   - Modified DiffusersFSDPEngine to support npu and optimized attention with _native_npu.
   - Extended vLLMOmniColocateWorkerExtension to include NPU sleep/wake-up logic using CaMemAllocator.
- Reward System:
   - Implemented DiffusionRewardLoopManager to handle non-tensor reward assembly, necessary for image-based RL feedback.
- Optimization:
   - Leveraged vllm-ascend's memory allocator to minimize fragmentation during heavy FSDP/Rollout/Reward model co-location.
- Docs & CI:
   - Added docs/start/flowgrpo_quickstart_npu.md for user onboarding.
   - Added tests/npu_smoke/ for offline/CI validation.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update the documentation.
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl-omni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...

### CODE REVIEW

File: docs/start/flowgrpo_quickstart_npu.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# Quickstart: FlowGRPO training on Qwen-Image OCR dataset with Ascend NPU
+
+Last updated: 05/09/2026
+
+## Introduction
+
+This guide launches FlowGRPO LoRA training for `Qwen-Image` OCR generation on Ascend NPU.
+
+## Environment preparation
```
Comment (zhtmike) [inline docs/start/flowgrpo_quickstart_npu.md]: Can we extract the installation related details under doc/start/, similar to install.md

File: verl_omni/reward_loop/reward_loop.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        from verl.utils.import_utils import load_extern_object
+
+        return load_extern_object(
+            module_path=reward_manager_cfg.module.path,
+            object_name=reward_manager_cfg.name,
+        )
+    raise ValueError(f"Unsupported reward manager source: {reward_manager_cfg.source}")
+
+
+class DiffusionRewardLoopManager(RewardLoopManager):
```
Comment (zhtmike) [inline verl_omni/reward_loop/reward_loop.py]: Currently we use the verl's RewardLoopManager directly. Any reason to create a new one?

File: verl_omni/reward_loop/reward_manager/visual.py [hunk 3]
```diff
@@ -16,12 +16,14 @@
...
 import torch
 from verl import DataProto
+from verl.experimental.reward_loop.reward_manager import register
 from verl.experimental.reward_loop.reward_manager.base import RewardManagerBase
 from verl.utils.reward_score import default_compute_score as _upstream_default_compute_score
 
 from verl_omni.utils.reward_score import default_compute_score_image
 
 
+@register("visual")
```
Comment (zhtmike) [inline verl_omni/reward_loop/reward_manager/visual.py]: We use the importlib method instead of registration, so RewardManager can be reused from verl

File: verl_omni/trainer/diffusion/ray_diffusion_trainer.py [hunk 4]
```diff
@@ -601,13 +601,13 @@ def init_workers(self):
             self.ref_policy_wg = self.actor_rollout_wg
 
         # create reward loop manager
-        from verl.experimental.reward_loop import RewardLoopManager
```
Comment (zhtmike) [inline verl_omni/trainer/diffusion/ray_diffusion_trainer.py:617]: Same

**Comment by zhtmike** [comment]
Thank you for your PR!
Overall looks good! Left with few comments

**Comment by zhtmike** [comment]
@knlnguyen1802 please take a look of vllm—omni related change. Thanks!

File: verl_omni/workers/rollout/vllm_rollout/utils.py [hunk 5]
```diff
@@ -45,6 +88,70 @@ def __new__(cls, **kwargs):
...
+        allocator = _get_npu_memory_allocator()
+        allocator.wake_up(tags=tags)
+
+        if len(self._sleep_saved_buffers) and self.model_runner is not None:
+            model = self.model_runner.pipeline
+            for name, buffer in model.named_buffers():
+                if name in self._sleep_saved_buffers:
+                    buffer.data.copy_(self._sleep_saved_buffers[name].data)
+            self._sleep_saved_buffers = {}
+        return True
```
Comment (knlnguyen1802) [inline verl_omni/workers/rollout/vllm_rollout/utils.py:153]: Prefer to move these change into vllm-omni

File: verl_omni/workers/engine/fsdp/diffusers_impl.py [hunk 6]
```diff
@@ -216,6 +216,9 @@ def _build_module(self):
             if self.model_config.enable_gradient_checkpointing:
                 module.enable_gradient_checkpointing()
 
+            if device_name == "npu" and hasattr(module, "set_attention_backend"):
```
Comment (zhtmike) [inline verl_omni/workers/engine/fsdp/diffusers_impl.py]: we already have `set_attention_backend` at line 211: `module.set_attention_backend(self.model_config.attn_backend)` . Pls merge them.

File: verl_omni/workers/rollout/vllm_rollout/utils.py [hunk 7]
```diff
@@ -45,6 +88,70 @@ def __new__(cls, **kwargs):
...
+            free_bytes_before_sleep = torch.npu.mem_get_info()[0]
+        except Exception:
+            pass
+
+        if level == 2 and self.model_runner is not None:
+            model = self.model_runner.pipeline
+            self._sleep_saved_buffers = {name: buffer.cpu().clone() for name, buffer in model.named_buffers()}
+
+        allocator = _get_npu_memory_allocator()
+        allocator.sleep(offload_tags=("weights",) if level == 1 else tuple())
```
[the remark this answers is not in the document]
Comment (panshaowu) [reply verl_omni/workers/rollout/vllm_rollout/utils.py:124]: I have double-checked the memory offloading logic in verl_omni/workers/rollout/vllm_rollout/utils.py. L119~L121 works when level == 2  and the current implementation is intentionally consistent with vLLMOmniColocateWorkerExtension. Our internal testing shows it works as expected for the current NPU
  integration. So it is unnecessary to revise this codes.

File: verl_omni/reward_loop/reward_loop.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+        if self.reward_model_manager is not None:
+            self.reward_model_manager.wake_up()
+
+        try:
+            chunks = data.chunk(len(self.reward_loop_workers))
+            outputs = ray.get(
+                [
+                    worker.compute_score_batch.remote(chunk)
+                    for worker, chunk in zip(self.reward_loop_workers, chunks, strict=True)
```
[the remark this answers is not in the document]
Comment (panshaowu) [reply verl_omni/reward_loop/reward_loop.py]: I have double-checked the the logic of DiffusionRewardLoopManager and RewardLoopManager. RewardLoopManager.compute_rm_score also sets strict=True. And I'm currently testing whether we can safely remove DiffusionRewardLoopManager and fallback to the base RewardLoopManager while ensuring NPU FlashAttention compatibility. So this code does not need to be modified for now.

File: examples/flowgrpo_trainer/run_qwen_image_ocr_lora_npu.sh
Comment (SamitHuang) [inline examples/flowgrpo_trainer/run_qwen_image_ocr_lora_npu.sh:1]: pls attach the wandb or tensorboard results in PR description

**Comment by zhtmike** [comment]
LGTM now

@SamitHuang @knlnguyen1802 any other comment?

**Review by knlnguyen1802 — APPROVED** [review]
LGTM now, we can move adapter code to vllm-omni later

**Review by zhtmike — APPROVED** [review]

File: verl_omni/reward_loop/reward_loop.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (panshaowu) [reply verl_omni/reward_loop/reward_loop.py]: DiffusionRewardLoopManager has been removed from this PR, so there is no need to discuss this question.

File: docs/start/flowgrpo_quickstart_npu.md (hunk 1 above)
Comment (panshaowu) [reply docs/start/flowgrpo_quickstart_npu.md]: Thanks for your suggestion. Content related to installing Ascend NPU dependencies have been moved into install.md.

File: verl_omni/reward_loop/reward_loop.py (hunk 2 above)
Comment (panshaowu) [reply verl_omni/reward_loop/reward_loop.py]: I created DiffusionRewardLoopManager to ensure NPU FlashAttention compatibility. But according to test results,  we can safely remove DiffusionRewardLoopManager and fallback to the base RewardLoopManager. So RewardLoopManager has been removed from codes of the PR.

File: verl_omni/reward_loop/reward_manager/visual.py (hunk 3 above)
Comment (panshaowu) [reply verl_omni/reward_loop/reward_manager/visual.py]: OK, related code has been modified.

File: verl_omni/trainer/diffusion/ray_diffusion_trainer.py (hunk 4 above)
Comment (panshaowu) [reply verl_omni/trainer/diffusion/ray_diffusion_trainer.py:617]: OK, DiffusionRewardLoopManager has been replaced with RewardLoopManager

**Comment by SamitHuang** [comment]
It's necessary to provide the training results for covergence reference

File: verl_omni/workers/engine/fsdp/diffusers_impl.py (hunk 6 above)
Comment (panshaowu) [reply verl_omni/workers/engine/fsdp/diffusers_impl.py]: OK, related code has been revised.

File: examples/flowgrpo_trainer/run_qwen_image_ocr_lora_npu.sh
Comment (panshaowu) [reply examples/flowgrpo_trainer/run_qwen_image_ocr_lora_npu.sh:1]: The long-term stability training experiment is in progress. It appears to be normal so far. I will supplement the training results after the experiment.

**Review by SamitHuang — APPROVED** [review]
LGTM. The reward curve looks increasing normally. please attach the reference performance in another PR.

File: examples/flowgrpo_trainer/run_qwen_image_ocr_lora_npu.sh
Comment (panshaowu) [reply examples/flowgrpo_trainer/run_qwen_image_ocr_lora_npu.sh:1]: The tensorboard results have been added to the PR description. The training results appears to meet expectations. And the convergence speed of NPUs is even faster than GPUs. We will try more experiments later.
